src/openms/source/METADATA/IdentificationData.cpp

```diff
@@ -0,0 +1,341 @@
+        for (ScoreList::const_reverse_iterator score_it = match.scores.rbegin();
+             score_it != match.scores.rend(); ++score_it)
+        {
+          const ScoreType& score = score_types.left.at(score_it->first);
+          if (score.params_key == step.params_key)
+          {
+            hit.setScore(score_it->second);
+            pair<DataQueryKey, ProcessingStepKey> key =
+              make_pair(query_key, *step_it);
+            psm_data[key].first.push_back(hit);
+            psm_data[key].second = score_it->first;
+            break;
+          }
+        }
+      }
+    }
+
+    set<ProcessingStepKey> steps;
+    for (auto psm_it = psm_data.begin(); psm_it != psm_data.end(); ++psm_it)
+    {
+      const DataQuery& query = data_queries.left.at(psm_it->first.first);
+      PeptideIdentification peptide;
+      static_cast<MetaInfoInterface&>(peptide) = query;
+      peptide.setRT(query.rt);
+      peptide.setMZ(query.mz);
+      peptide.setMetaValue("spectrum_reference", query.data_id);
+      peptide.setHits(psm_it->second.first);
+      const ScoreType& score_type = score_types.left.at(psm_it->second.second);
+      peptide.setScoreType(score_type.name);
+      peptide.sortByRank();
+      peptide.setIdentifier(String(psm_it->first.second));
+      peptides.push_back(peptide);
+      steps.insert(psm_it->first.second);
+    }
+
+    map<ProcessingStepKey, pair<vector<ProteinHit>, ScoreTypeKey>> prot_data;
+    for (unordered_map<ParentMoleculeKey, ParentMetaData>::const_iterator
+           par_it = parent_meta_data.begin(); par_it != parent_meta_data.end();
+         ++par_it)
+    {
+      if (par_it->second.molecule_type != MT_PROTEIN) continue;
+      const ParentMetaData& meta = par_it->second;
+      ProteinHit hit;
+      hit.setSequence(meta.sequence);
+      hit.setDescription(meta.description);
+      hit.setCoverage(meta.coverage);
+      hit.setAccession(parent_molecules.left.at(par_it->first));
+      static_cast<MetaInfoInterface&>(hit) = meta;
+      // find all steps that assigned a score:
+      for (vector<ProcessingStepKey>::const_iterator step_it =
+             meta.processing_steps.begin(); step_it !=
+             meta.processing_steps.end(); ++step_it)
+      {
+        const DataProcessingStep& step = processing_steps.at(*step_it);
+        // give priority to "later" scores:
+        for (ScoreList::const_reverse_iterator score_it = meta.scores.rbegin();
+             score_it != meta.scores.rend(); ++score_it)
+        {
+          const ScoreType& score = score_types.left.at(score_it->first);
+          if (score.params_key == step.params_key)
+          {
+            hit.setScore(score_it->second);
+            prot_data[*step_it].first.push_back(hit);
+            prot_data[*step_it].second = score_it->first;
+            steps.insert(*step_it);
+            break;
+          }
+        }
+      }
+    }
+
+    for (set<ProcessingStepKey>::const_iterator step_it = steps.begin();
```

Comment:
consider using range iteration:
 for (auto const & s : steps) 
it makes refactoring easier (and is also easier to read)

Reply:
I'm not used to that yet. ;-)